### ISSUE TITLE
Container lifecycle hook support timeout limit

### ISSUE THREAD

**Issue by qingwave** [kubernetes/kubernetes#99761]
#### What would you like to be added:
Now, container lifecycle `postStart` or `preStop` just a [handler](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#lifecycle-v1-core), timeout should be supported to aviod the hook run long time. 
```
type Handler struct {
	// One and only one of the following should be specified.
	// Exec specifies the action to take.
	// +optional
	Exec *ExecAction `json:"exec,omitempty" protobuf:"bytes,1,opt,name=exec"`
	// HTTPGet specifies the http request to perform.
	// +optional
	HTTPGet *HTTPGetAction `json:"httpGet,omitempty" protobuf:"bytes,2,opt,name=httpGet"`
	// TCPSocket specifies an action involving a TCP port.
	// TCP hooks not yet supported
	// TODO: implement a realistic TCP lifecycle hook
	// +optional
	TCPSocket *TCPSocketAction `json:"tcpSocket,omitempty" protobuf:"bytes,3,opt,name=tcpSocket"`
}
```

#### Why is this needed:
The `PostStart` hook may block forever, as decribe in #68013, timeout should be supported to aviod it.

**Comment by qingwave** [comment]
sig/node
kind/feature

**Comment by benzvan** [comment]
I see this never got addressed and is closed but I thought I'd chime in on what I believe to be the existing solution here. The `terminationGracePeriodSeconds` timer starts right before the preStop hook is executed. That timer is 30 seconds by default, so if your preStop hook fails to finish in 30 seconds, the container will be sent a SIGKILL and deleted. If your preStop hook finishes in less than 30 seconds, the container is sent SIGTERM and the timer continues. 

https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination

**Comment by TjlHope** [comment]
Sorry for the necro, but JIC someone else hits this issue I've worked around it by pre-chaining my `postStart` hooks with a `timeout` command (I've made sure all our containers have `busybox` in them), e.g. 
```
  lifecycle:
    postStart:
      exec: ["timeout", "60", "hook-that-might-hang"]
```

This way if the real hook doesn't return within the specified timeout it'll get sent a `TERM`.

In reality I've wrapped the logic in a `postStart_timeout` script that will also chain in `dumb-init` (if available) as `timeout` will only signal its immediate child, not the whole session (process tree).

Note: this is only needed for `postStart`, as mentioned `terminationGracePeriodSeconds` handles the `preStop` case nicely.
